[FIXES] Issue #13486 handle fillstate

arpit551 opened the pull request:
Fixes [issue](https://github.com/zulip/zulip/issues/13486)

Currently tested manually. 

Commits have a detailed message about what has been done.

@timabbott kindly review this PR so that I can look forward to write test cases

timabbott commented:
@arpit551 thanks for working on this!

I think it'd be helpful to focus on just the "add realm argument to analytics" commit and tests for now; that alone is a fairly complex migration (it might be best for you to pull the import/export changes off this PR to simplify the discussion).  Looking at that commit, you have a lot of seemingly accidental changes to whitespace/style that need to be removed for this to be reviewable; did your editor add those automatically?

Check out the Zulip commit guidelines for more details: https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages

arpit551 commented:
@timabbott actually I must have used vs code PEP formatter. I have cleared the changes made by it. and also removed the commits of import and export. 
Added some tests too. 
Furthur if we want to use count stats for single or specified realms (other than for import-export) then I think we have to add realm argument in Fill_State table and then count stats for each realm, though it may be time-consuming but will be helpful for cases where stats for all realms are not needed.

mateuszmandera commented:
You have mypy (typing) error.
```
class DataCollector:
    def __init__(self, output_table: Type[BaseCount],
                 pull_function: Optional[Callable[[str, datetime, datetime], int]]) -> None:
        self.output_table = output_table
        self.pull_function = pull_function
```
pull_function is annotated as a function taking 3 arguments, but you're calling it with 4 on line 136:
```
rows_added = stat.data_collector.pull_function(stat.property, start_time, end_time, realm)
```

arpit551 commented:
Fixed it and this is ready for review.

timabbott commented on analytics/lib/counts.py:
```diff
@@ -79,7 +79,8 @@ def process_count_stat(stat: CountStat, fill_to_time: datetime) -> None:
     if floor_to_hour(fill_to_time) != fill_to_time:
-        raise ValueError("fill_to_time must be on an hour boundary: %s" % (fill_to_time,))
+        raise ValueError(
+            "fill_to_time must be on an hour boundary: %s" % (fill_to_time,))
```
Can you remove these spurious reformattings?  They make reading your changes difficult.

timabbott commented on analytics/lib/counts.py:
```diff
@@ -113,7 +115,7 @@ def process_count_stat(stat: CountStat, fill_to_time: datetime) -> None:
         start = time.time()
         do_update_fill_state(fill_state, currently_filled, FillState.STARTED)
-        do_fill_count_stat_at_hour(stat, currently_filled)
+        do_fill_count_stat_at_hour(stat, currently_filled, realm)
```
So, the code as of this commit is such that it deserves a comment that the code is in an intermediate state where we're only updating the count_stat for a single realm while updating the global FillState, that this will be fixed in future commits, and until then the `realm` argument is here just for plumbing, not for production use.

arpit551 replied:
I think that's what I want to get cleared as I mentioned in [comment](https://github.com/zulip/zulip/pull/13608#issuecomment-572023210). I wanted to know what should be the correct way handling global FillState should we add a realm argument there or there is a better approach than it. For now, I will comment it there.

timabbott replied:
I think we're going to write future commits later in this effort that will adjust the model to support a per-realm FillState. 
 We haven't designed the model for that yet, so for now a comment reminding us that we'll need to do something there is the best we can do.

timabbott replied:
(Or we might go with a solution that just involves skipping this step when we have a realm specified?  Again, I'm not sure, and want to defer designing that until we have this refactoring merged).

timabbott commented on analytics/lib/counts.py:
```diff
@@ -147,11 +149,17 @@ def do_delete_counts_at_hour(stat: CountStat, end_time: datetime) -> None:
     # Aggregate into RealmCount
     output_table = stat.data_collector.output_table
+    if realm is not None:
+        realm_clause = "AND zerver_realm.id = " + str(realm.id)
```
Let's use a `%` comprehension here; that's clearer than adding a `str()`.

timabbott commented on analytics/lib/counts.py:
```diff
@@ -237,33 +247,40 @@ def do_drop_single_stat(property: str) -> None:
+    if realm is not None:
+        realm_id = realm.id
+    else:
+        realm_id = ""
```
I think it'd be more correct to only pass a `realm_id` argument in the case where `realm is not `None`, by constructing an argument dict and mutating it before passing that to `cursor.execute`.  Otherwise, if `realm is None`, we're passing invalid input.

timabbott commented on analytics/lib/counts.py:
```diff
@@ -273,17 +290,23 @@ def do_pull_minutes_active(property: str, start_time: datetime, end_time: dateti
     seconds_active = defaultdict(float)  # type: Dict[Tuple[int, int], float]
     for user_id, realm_id, interval_start, interval_end in user_activity_intervals:
+        if realm is not None and realm.id is not realm_id:
+            continue
```
`realm.id != realm_id` is the normal way to write this.

timabbott commented on analytics/lib/counts.py:
```diff
@@ -273,17 +290,23 @@ def do_pull_minutes_active(property: str, start_time: datetime, end_time: dateti
         start = max(start_time, interval_start)
         end = min(end_time, interval_end)
         seconds_active[(user_id, realm_id)] += (end - start).total_seconds()
-
```
Can you add back this blanik line?  It was there for readability.

timabbott commented:
Thanks for working on this @arpit551!  I posted a batch of comments.

timabbott commented on analytics/lib/counts.py:
```diff
@@ -425,11 +479,17 @@ def do_pull_minutes_active(property: str, start_time: datetime, end_time: dateti
+        realm_clause = ""
+    else:
+        realm_clause = "realm_id = %%(realm_id)s AND"
+    return """
```
One other question I had is how much you've thought about the API detail of whether we have the `_query` functions insert the `realm_id` themselves or just have `realm_clause = "realm_id = %(realm_id)s AND" % (realm.id,)` in this case?  I think the latter would result in cleaner code (since we already passed the `Realm` object here, there's no reason to not do the substitution now before we pass it into SQL).

arpit551 replied:
@timabbott I think  realm_clause = "realm_id = %(realm_id)s AND" % (realm.id,) would be better as it will result in cleaner code as well it makes us bound to pass realm in the other direction too i.e other than COUNT_STATS we have to pass realm in do_fill_count_stat_at_hour which I suppose is a good practice as we realm is required there. Though I am confused about whether to pass realm_id with string replacement or by giving it as a parameter to cursor.execute()

timabbott replied:
Yeah, I'd appreciate it if you made that change.  I think either `count_message_type_by_user_query` should take a boolean (for whether there's a realm) or it should take an `Optional[realm]` and do the realm-level filtering itself (rather than expecting the resulting query to have a realm passed in).  

The latter option (i.e. what I proposed above) seems cleaner to me; can you make that change?

arpit551 commented:
@timabbott done the requested changes.

arpit551 commented:
@timabbott I have pushed the requested changes.

timabbott commented on analytics/lib/counts.py:
```diff
@@ -471,71 +530,88 @@ def do_pull_minutes_active(property: str, start_time: datetime, end_time: dateti
+
+def get_count_stats(realm: Optional[Realm]=None) -> Dict[str, CountStat]:
+    count_stats_ = get_count_stats_(realm)
+    return OrderedDict([(stat.property, stat) for stat in count_stats_])
```
Why do you need the `get_count_stats_` function, rather than e.g. having `get_count_stats` just return an OrderedDict if all the callers want that format?

arpit551 replied:
@timabbott this was due to requirement for the list of COUNT_STATS objects in future but its highly unlikely that it will be needed. So pushed the cleaner code.

timabbott replied:
Yeah, I think making code uglier to simplify support for an unlikely potential future use case is a pattern we want to avoid wherever we can.

arpit551 commented:
@timabbott  should I add the refactoring and import-export commit now? for the review

timabbott commented:
I think we should plan to do those on a new PR; this one has enough clutter already.

timabbott commented:
I edited the title for that new PR to be more clear about what's it about.

timabbott commented:
@arpit551 thanks for all your work on this; it's a pretty big change.  I merged this as b23a5431cd949ad5a2834a4905eacfb2533bbcff after a few key changes:
* Changed the aggregation logic to not update `InstallationCount` if `realm is not None`; this is I think more consistent with how we'll want this to work in the near term.  Later, we may need to rewrite the InstallationCount rows entirely after a RealmCount update with a realm parameter (using data from all realms), but I don't want to figure that out yet.
* Moved around the comments documenting how the `realm` argument works to be at the start of `process_count_stat`, since that's where one will look to understand how this works.

I have a few follow-up questions about how this works that I'd love your help in resolving but I think are best fixed forward (so I don't need to re-review the 450 lines of this already merged):
* A couple CountStat objects, specifically `invites_sent::day` and `active_users_log:is_bot:day` and `minutes_active::day` don't seem to have a realm argument; is that expected?  We might want to document what's going on with comments.  
* Does `pull_function` actually need a `realm` argument?  It seems to not use it.

If there are changes to make for these, put them either in a new PR or as commits rebased to the start of your other branch.

Thanks for all your work on this thus far @arpit551!  This PR became massively easier to understand and merge with the changes you made in review.

arpit551 commented:
@timabbott The two stats `invites_sent::day` and `active_users_log:is_bot:day` are logging stat so in there pull_functions are not present so realm argument can only be used in do_aggregate_to_summary_table function for them.
Regarding the second question, we need realm argument in pull function because `minutes_active::day` stat, uses do_pull_minutes_active as pull function and there we need realm.
So we have to pass realm argument in all pull functions.

timabbott commented:
OK, great; can you open a small PR (or add a commit at the start of your other PR, may be simpler) adding comments explaining those details in appropriate places?  That'll help a lot for making the code understandable for future developers.

arpit551 commented:
@timabbott  Opened a separate PR for it.
